elements/pl-units-input/pl-units-input.py

```diff
@@ -0,0 +1,462 @@
+            if submitted_answer is not None:
+                html_params["raw_submitted_answer"] = pl.escape_unicode_string(
+                    submitted_answer
+                )
+
+        if score is not None:
+            score_type, score_value = pl.determine_score_params(score)
+            html_params[score_type] = score_value
+
+        feedback = partial_scores.get("feedback")
+        show_feedback = pl.get_boolean_attrib(
+            element, "show-feedback", SHOW_FEEDBACK_DEFAULT
+        )
+        if feedback is not None and show_feedback:
+            html_params["feedback"] = feedback
+
+        html_params["error"] = html_params["parse_error"] or html_params.get(
+            "missing_input", False
+        )
+        with open(UNITS_INPUT_MUSTACHE_TEMPLATE_NAME, "r", encoding="utf-8") as f:
+            return chevron.render(f, html_params).strip()
+
+    # TODO make this display consistent with number input (sigfigs and such)
+    elif data["panel"] == "answer":
+        a_tru = data["correct_answers"].get(name, None)
+
+        if a_tru is None:
+            return ""
+
+        ureg = UnitRegistry(cache_folder=":auto:")
+        a_tru_parsed = ureg.Quantity(a_tru)
+
+        if grading_mode is GradingMode.UNITS_ONLY:
+            a_tru_string = f"{a_tru_parsed.units}"
+        else:
+            a_tru_string = f"{a_tru_parsed}"
+
+        html_params = {
+            "answer": True,
+            "label": label,
+            "a_tru": a_tru_string,
+            "suffix": suffix,
+        }
+        with open(UNITS_INPUT_MUSTACHE_TEMPLATE_NAME, "r", encoding="utf-8") as f:
+            return chevron.render(f, html_params).strip()
+
+    assert_never(data["panel"])
+
+
+def parse(element_html: str, data: pl.QuestionData) -> None:
+    element = lxml.html.fragment_fromstring(element_html)
+    name = pl.get_string_attrib(element, "answers-name")
+    allow_blank = pl.get_string_attrib(element, "allow-blank", ALLOW_BLANK_DEFAULT)
+    blank_value = pl.get_string_attrib(element, "blank-value", BLANK_VALUE_DEFAULT)
+
+    units_only = (
+        pl.get_enum_attrib(element, "grading-mode", GradingMode, GRADING_MODE_DEFAULT)
+        is GradingMode.UNITS_ONLY
+    )
+
+    # retrieves submitted answer
+    a_sub = data["submitted_answers"].get(name, None)
+    if a_sub is None:
+        data["format_errors"][name] = "No submitted answer."
+        data["submitted_answers"][name] = None
+        return
+
+    # checks for blank answer
+    if not a_sub:
+        if allow_blank:
+            data["submitted_answers"][name] = blank_value
+        else:
+            data["format_errors"][
+                name
+            ] = "Invalid format. The submitted answer was left blank."
+            data["submitted_answers"][name] = None
+
+        return
+
+    # TODO double check that doing this is ok. May need to switch to directory local to question / server
+    ureg = UnitRegistry(cache_folder=":auto:")
```

Comment:
@nwalters512 I think these cache files are the last real item that might be of concern in this PR (other pending changes are minor cosmetic). There's documentation about what's going on in this cache folder here: https://pint.readthedocs.io/en/develop/advanced/performance.html#speed-up-registry-instantiation

Should this be made local to the assessment? Not sure if using the default across courses could create potential issues, or whether old cache files need to be cleaned out. Maybe there's a way to store them in the DB local to the question? Without setting up a cache like this, this element is really really slow.

Reply:
This should specifically be in `/tmp/`, as that's the only directory guaranteed to be writable in prod. There's precedent for this with the `matplotlib` cache:

https://github.com/PrairieLearn/PrairieLearn/blob/4b07deac3d386a1843a1b2423df07be1c0dd9562/python/zygote.py#L30-L38

If this is truly really slow to load, it should be preloaded in the zygote that we fork from:

https://github.com/PrairieLearn/PrairieLearn/blob/4b07deac3d386a1843a1b2423df07be1c0dd9562/python/zygote.py#L46-L59

Reply:
Gotcha. Using the a directory in `/tmp/` should be fine. The issue isn't that the module is slow to load (though it's probably also worth preloading in `zygote.py`), the issue is that the `UnitRegistry` object is expensive to construct, so using a cached version pickles the object for subsequent constructions. But if there's a folder where that caching can happen, we shouldn't run into any issues.